File: statistics/C/src/List.c
/*
* @Author: Jingyuexing
* @Date:   2020-12-09 19:19:21
* @Last Modified by:   Jingyuexing
* @Last Modified time: 2020-12-09 20:21:46
*/
#include "List.h"

ListNode *init(double element){
    ListNode *newNode = (ListNode *)malloc(sizeof(ListNode));
    newNode->next = NULL;
    newNode->element = element;
    return newNode;
}

void remove(){

}
void insert(List l,int index,double element){

}
void append(List l,double element){
    l.postion->next = init(element);
    l.postion = l.postion->next;
};
int isEmpty(List l){
    return l.head->next == NULL;
}

/**
 * 创建一个List
 * @param  element [description]
 * @return         [description]
 */
List newList(double element){
    auto List *newList = (List *)malloc(sizeof(List));
    (*newList).head = init(0x000);
    (*newList).head->next = init(element);
    (*newList).self = &newList;
    (*newList).head = NULL;
    return *newList;
}
void deleteList(List l){
    free(&l);
}
ListNode *indexOf(List l,int index){
    ListNode *target = NULL;
    target = l.head->next;
    while (target!=NULL && target->index != index) {
        target = target->next;
    }
    return target;
}

int main(int argc, char const *argv[])
{
    List myList;
    myList.head = init(0x000);
    ListNode *currNode;
    currNode = myList.head->next;
    int i =0;
    while (currNode != NULL) {
        currNode->index = i;
        i++;
    }
    return 0;
}

File: statistics/Lua/src/List.lua
-- @Author: admin
-- @Date:   2021-12-12 09:31:48
-- @Last Modified by:   admin
-- @Last Modified time: 2021-12-23 03:51:12



local List = {
    head=nil,
    pos=nil,
    length=0
}
local NULL = {}
local Node = {
    element= nil,
    next = NULL
}
function Node.new(element)
    local node = {
    }
    setmetatable(node,{__index=Node})
    node.element = element
    node.next = NULL
    return node
end


function List.new()
    local self = {}
    setmetatable(self,{__index=List})
    return self
end

function List.append(self,element)
    local Node = Node.new(element)
    if self.head ~= nil then
        self.pos.next = Node
        self.pos = self.pos.next
    else
        self.head = {}
        self.pos = {}
        self.head = Node
        self.pos = self.head
    end
    self.length = self.length+1
end

function List.find(self,element)
    local current = self.head
    while current.element ~= element do
        current = current.next
    end
    return current
end

function List.remove(self,element)
    local targetNode = self.find(self,element)
    targetNode.element = targetNode.next.element
    local nextNode = targetNode.next
    targetNode.next = nextNode.next
    nextNode=nil
    return self
end

return List
